masu/processor/providers/aws/report_downloader.py

```diff
@@ -0,0 +1,208 @@
+#
+# Copyright 2018 Red Hat, Inc.
+#
+# This program is free software: you can redistribute it and/or modify
+# it under the terms of the GNU Affero General Public License as
+# published by the Free Software Foundation, either version 3 of the
+# License, or (at your option) any later version.
+#
+# This program is distributed in the hope that it will be useful,
+# but WITHOUT ANY WARRANTY; without even the implied warranty of
+# MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
+# GNU Affero General Public License for more details.
+#
+# You should have received a copy of the GNU Affero General Public License
+# along with this program.  If not, see <https://www.gnu.org/licenses/>.
+#
+"""AWS CUR Downloader."""
+
+import json
+import os
+from datetime import datetime
+
+import boto3
+from dateutil.relativedelta import relativedelta
+
+from masu.processor.exceptions import MasuConfigurationError
+from masu.processor.report_downloader import ReportDownloader
+from masu.providers import DATA_DIR
+from masu.providers.aws.sts import get_assume_role_session
+
+
+class AWSReportDownloader(ReportDownloader):
+    """
+    AWS Cost and Usage Report Downloader.
+
+    For configuration of AWS, see
+    https://docs.aws.amazon.com/awsaccountbilling/latest/aboutv2/billing-reports-costusage.html
+    """
+
+    def __init__(self, **kwargs):
+        """
+        Constructor.
+
+        Args:
+            customer     (CustomerDBAccessor) the customer accessor
+            provider     (ProviderAuthDBAccessor) the provider accessor
+            report_stats (ReportStatsDBAccessor) the report stats accessor
+
+        """
+        try:
+            assert 'provider' in kwargs
+            assert 'customer' in kwargs
+            assert 'report_stats' in kwargs
+        except AssertionError:
+            raise MasuConfigurationError('Missing required argument.')
+
+        super().__init__(**kwargs)
+
+        self.provider = kwargs.get('provider')
+        self.customer = kwargs.get('customer')
+        self.report_stats = kwargs.get('report_stats')
+
+        self.customer_schema = self.customer.get_schema_name()
+
+        self.logger.debug('Connecting to AWS...')
+        session = get_assume_role_session(
```

Comment:
+1 agree. I haven't handled those exceptions yet, primarily because I hadn't fully decided _where_ to handle them yet.

I was debating whether to catch the exceptions at the caller and re-raise when needed versus catching the exception later on. So, I left it alone until I needed to become more opinionated on where to catch the thrown exceptions.

Reply:
I think it could make sense to catch the exceptions here, log it, and then raise an AWSReportDownloader exception.  That way the caller of AWSReportDownloader will know at instantiation time that the object is not usable and it could handle things.

This will be the orchestrator that will have to deal with a useless downloader and it would have to signal to the service admin some how some explanation of whats wrong (invalid credentials, networking problems, etc).

The orchestrator integration will come next but for now I think it's good to have the downloader absorb and interpret the boto3 exception and raise it's own.

Reply:
By the way I'm good with leaving this exception work to be done when we integrate with the orchestrator.  Just leaving some thoughts here.